### PR TITLE
bugzilla: restrict automation for bug in groups

### PR DESCRIPTION
**Pull request by AlexNPavel**
This PR adds a new `allowed_groups` config to the bugzilla plugin
configuration and changes the behavior of the plugin to ignore bugs
that are part of groups not in the `allowed_groups` config.

/cc @stevekuznetsov

### CODE REVIEW

File: prow/plugins/bugzilla/bugzilla.go [hunk 1]
```diff
@@ -438,6 +438,17 @@ func processQuery(query *emailToLoginQuery, email string, log *logrus.Entry) str
...
 func handle(e event, gc githubClient, bc bugzilla.Client, options plugins.BugzillaBranchOptions, log *logrus.Entry) error {
 	comment := e.comment(gc)
+	// check if bug is part of a restricted group
+	if !e.missing {
+		bug, err := getBug(bc, e.bugId, log, comment)
+		if err != nil || bug == nil {
+			return err
+		}
+		if !isBugAllowed(bug, options.AllowedGroups) {
+			// ignore bugs that are in non-allowed groups for this repo
```
Comment (stevekuznetsov) [inline prow/plugins/bugzilla/bugzilla.go:448]: We don't have to leak which group the bug is in, but we should comment for the user to know that the bug is not in the allowed groups list, and list which groups are allowed. That way they can see feedback and we leak nothing.

File: prow/plugins/bugzilla/bugzilla_test.go [hunk 2]
```diff
@@ -1121,8 +1121,8 @@ Instructions for interacting with me using PR comments are available [here](http
 			}},
 			prs:     []github.PullRequest{{Number: base.number, Merged: true}},
 			options: plugins.BugzillaBranchOptions{StateAfterMerge: &modified}, // no requirements --> always valid
-			expectedComment: `org/repo#1:@user: An error was encountered searching for external tracker bugs for bug 123 on the Bugzilla server at www.bugzilla:
-> injected error adding external bug to bug
```
Comment (stevekuznetsov) [inline prow/plugins/bugzilla/bugzilla_test.go:1125]: Why did unrelated tests change?

Comment (AlexNPavel) [reply prow/plugins/bugzilla/bugzilla_test.go:1125]: Because the fake client is extremely non-specific with injected errors. We simply mark a bug to inject errors for and then that error occurs for any operation (except create bug). Since the first thing we do in handle now is get the bug, we hit the "getting bug" error before the error for adding an external bug, which used to be the first error we would hit in this unit test.

We would have to change the way the fake client injects errors and make it more specific/targeted based on operation to make prevent this test from changing.

File: prow/plugins/bugzilla/bugzilla.go (hunk 1 above)
Comment (AlexNPavel) [reply prow/plugins/bugzilla/bugzilla.go:448]: This would cause us to comment every time there is a change though. Would it make sense to only have it respond about the bug group if it was triggered by a `/bugzilla refresh` and ignore in all other cases?

Comment (stevekuznetsov) [reply prow/plugins/bugzilla/bugzilla.go:448]: Yeah, that sounds good

**Comment by stevekuznetsov** [comment]
@petr-muller however, a deny-list allows leaks and would require that the sensitive names be listed in configuration. An allow-list is a fail-safe.

**Comment by petr-muller** [comment]
> @petr-muller however, a deny-list allows leaks and would require that the sensitive names be listed in configuration. An allow-list is a fail-safe.

@stevekuznetsov Okay. Out of curiosity, how would I express "no group"? Or is it not a thing in BZ?

**Comment by AlexNPavel** [comment]
> @stevekuznetsov Okay. Out of curiosity, how would I express "no group"? Or is it not a thing in BZ?

By default, bugzilla bugs have no group (unless they are marked as private, which sets the group to "redhat" for redhat bugzilla) and I believe they can only be added to a group after they are created. Adding groups restricts which users are allowed to view the bug. Bugs with no group are considered public.

**Comment by petr-muller** [comment]
@AlexNPavel that's what I expected, and that's why I'm wondering. Are we sure that bugs that need to be linked somewhere are always in a group? Sounds like many of them will not have a group, and with allowlist we have no way to allow such bugs to be linked...

**Comment by AlexNPavel** [comment]
> @AlexNPavel that's what I expected, and that's why I'm wondering. Are we sure that bugs that need to be linked somewhere are always in a group? Sounds like many of them will not have a group, and with allowlist we have no way to allow such bugs to be linked...

This change doesn't block bugs that aren't in a group. It only blocks bugs if they are part of a group that's not allowed. The `isBugAllowed` function returns true for bugs with no groups.
